### PR TITLE
[OAuth] Scope to list datasets

### PR DESCRIPTION
**Pull request by gonzaloriestra**
Closes https://github.com/CartoDB/cartodb/issues/15012

The final name to request the permission is `datasets:metadata`. There's no need to grant permissions or add roles, as we can just check if this scope is present and use the master API key to retrieve the list.

The grant schema will be like this for now:
```
{
  type: 'database',
  table_metadata: []
}
```

This way, we could add later more conditions like schema or dataset name, if needed.

I've also made a small refactor to move the classes in Scopes in different files.

### CODE REVIEW

**Comment by gonzaloriestra** [comment]
retest this please

File: lib/formats/carto/api_key/grants.json [hunk 1]
```diff
@@ -111,6 +111,9 @@
             "profile"
           ]
         }
+      },
+      "table_metadata": {
+        "type": "array"
```
Comment (ethervoid) [inline lib/formats/carto/api_key/grants.json]: Why is type array here? could be aboolean for example

File: lib/carto/oauth_provider/scopes/category.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+module Carto
```
Comment (ethervoid) [inline lib/carto/oauth_provider/scopes/category.rb]: :heart: the refactor but it should've been done in other PR apart

File: spec/lib/carto/oauth_provider/scopes/scopes_spec.rb [hunk 3]
```diff
@@ -601,5 +603,24 @@
       scopes_by_category = Carto::OauthProvider::Scopes.scopes_by_category(create_scope, [])
       expect(scopes_by_category).to(eq(expected))
     end
+
+    it 'shows create table permission' do
```
Comment (ethervoid) [inline spec/lib/carto/oauth_provider/scopes/scopes_spec.rb]: Should be named `shows list datasets permission`

File: spec/requests/carto/oauth_provider_controller_spec.rb [hunk 4]
```diff
@@ -112,15 +112,6 @@ def validate_token_response(parameters, access_token, refresh_token = nil)
         expect(qs['error']).to(eq('invalid_scope'))
       end
 
-      it 'redirects with an error if requesting non-existent schemas' do
```
Comment (ethervoid) [inline spec/requests/carto/oauth_provider_controller_spec.rb]: Why did you remove these tests?

File: spec/models/carto/api_key_spec.rb [hunk 5]
```diff
@@ -456,6 +456,17 @@ def validate_view_api_key(view_name, create_query, drop_query)
         }.to_not raise_error
       end
 
+      it 'validates tables_metadata grant' do
```
Comment (ethervoid) [inline spec/models/carto/api_key_spec.rb]: I miss a test where you actually ask for the metadata using an api key with the grant

File: lib/formats/carto/api_key/grants.json (hunk 1 above)
Comment (gonzaloriestra) [reply lib/formats/carto/api_key/grants.json]: I added it as an array to follow the same format as other properties, and also because later we may want to allow finer-grained scopes, like restricting by schema or datasets. And if we finally do it, it would break compatibility. What do you think?

File: lib/carto/oauth_provider/scopes/category.rb (hunk 2 above)
Comment (gonzaloriestra) [reply lib/carto/oauth_provider/scopes/category.rb]: Sorry, you are right, this is too big...

File: spec/requests/carto/oauth_provider_controller_spec.rb (hunk 4 above)
Comment (gonzaloriestra) [reply spec/requests/carto/oauth_provider_controller_spec.rb]: They were duplicated (just above).

File: spec/models/carto/api_key_spec.rb (hunk 5 above)
Comment (gonzaloriestra) [reply spec/models/carto/api_key_spec.rb]: I've added some tests for that in the next PR (there's no code to get the metadata here).

File: lib/formats/carto/api_key/grants.json (hunk 1 above)
Comment (ethervoid) [reply lib/formats/carto/api_key/grants.json]: It's a bit weird to see always an empty array but ok

File: spec/models/carto/api_key_spec.rb (hunk 5 above)
Comment (ethervoid) [reply spec/models/carto/api_key_spec.rb]: Ok

**Comment by alrocar** [comment]
Acceptance 🍏
